spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/source/StructInternalRow.java

```diff
@@ -338,11 +348,37 @@ private ArrayData collectionToArrayData(Type elementType, Collection<?> values)
             array ->
                 (BiConsumer<Integer, Map<?, ?>>)
                     (pos, map) -> array[pos] = mapToMapData(elementType.asMapType(), map));
+      case VARIANT:
+        return fillArray(
+            values,
+            array -> (BiConsumer<Integer, Object>) (pos, v) -> array[pos] = toVariantVal(v));
       default:
         throw new UnsupportedOperationException("Unsupported array element type: " + elementType);
     }
   }
 
+  private static VariantVal toVariantVal(Object value) {
+    if (value instanceof VariantVal) {
+      return (VariantVal) value;
```

Comment:
Can you help me understand what kind of data will be present ? From other types such as Map, Struct, seems we should only expect Variant here and then we are converting to VariantVal. 

By checking the usage, seems we should expect Variant and we should not see VariantVal?

Reply:
@aokolnychyi  Can you help take a look? I think we can remove.

Reply:
I removed this conditional branch.